src/ucp/proto/proto.c

```diff
@@ -54,29 +54,33 @@
     _macro(ucp_rndv_rkey_ptr_proto) \
     _macro(ucp_am_eager_short_proto) \
     _macro(ucp_tag_offload_eager_bcopy_single_proto) \
-    UCP_AMO_PROTO_LIST(_macro, post) \
-    UCP_AMO_PROTO_LIST(_macro, fetch) \
-    UCP_AMO_PROTO_LIST(_macro, cswap)
+    UCP_PROTO_AMO_FOR_EACH(_macro, post) \
+    UCP_PROTO_AMO_FOR_EACH(_macro, fetch) \
+    UCP_PROTO_AMO_FOR_EACH(_macro, cswap)
 
 #define UCP_PROTO_DECL(_proto) extern ucp_proto_t _proto;
 
-#define UCP_PROTO_DEF(_proto) &_proto,
-
+#define UCP_PROTO_ENTRY(_proto) &_proto,
 
 /* Declare all proto objects */
-UCP_PROTO_LIST(UCP_PROTO_DECL)
-
+UCP_PROTO_FOR_EACH(UCP_PROTO_DECL)
 
 const ucp_proto_t *ucp_protocols[] = {
-    UCP_PROTO_LIST(UCP_PROTO_DEF)
+    UCP_PROTO_FOR_EACH(UCP_PROTO_ENTRY)
 };
-unsigned ucp_protocols_count       = ucs_static_array_size(ucp_protocols);
 
 const char *ucp_proto_perf_types[] = {
     [UCP_PROTO_PERF_TYPE_SINGLE] = "single",
     [UCP_PROTO_PERF_TYPE_MULTI]  = "multi"
 };
 
+unsigned ucp_protocols_count(void)
+{
+    UCS_STATIC_ASSERT(ucs_static_array_size(ucp_protocols) <
+                      UCP_PROTO_MAX_COUNT);
```

Comment:
1. can we remove UCP_PROTO_MAX_COUNT macro altogether?
2. maybe assert ucs_static_array_size(ucp_protocols) < 64 because of the bitmap usage?
3. maybe use helper var (size_t num_protocols = ucs_static_array_size(ucp_protocols) )

Reply:
right.. so better revert to static assert and no helper var